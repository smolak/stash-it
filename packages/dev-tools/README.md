# `@stash-it/dev-tools`

![logo-stash-it-color-dark 2x](https://user-images.githubusercontent.com/1819138/30385483-99fd209c-98a7-11e7-85e2-595791d8d894.png)

Set of tools for development purposes.

## Table of Contents

- [Installation](#installation)
- [Usage](#usage)
- [License](#license)
- [Contribution](#contribution)

npm
```bash
npm install @stash-it/dev-tools
```

deno
```bash
deno add @stash-it/dev-tools
```

yarn
```bash
yarn dlx jsr add @stash-it/dev-tools
```

pnpm
```bash
pnpm dlx jsr add @stash-it/dev-tools
```

bun
```bash
bunx jsr add @stash-it/dev-tools
```

## Usage

### `runAdapterTests`

If you want to create an adapter, it should pass all of those tests. Here's how you can use them.
Let's say you use vitest:

```typescript
// Vitest stuff
import { describe } from "vitest";

import { adapterTestCases } from '@stash-it/dev-tools';

// Your adapter.
import { YourAdapter } from './your-adapter';

describe('YourAdapter', () => {
  const adapter = new YourAdapter();
  
  // Add whatever validation or other stuff you need.
  
  // Run the tests.
  adapterTestCases(adapter);
});
```

This helper function accepts an adapter instance and runs tests on it. It will throw an error if any of the tests fail.
It is also capable of cleaning up after self. Meaning, if you're operating on a persistent storage, it will delete all of the
items created during the tests.

### `getHandler`

If you want to creat a plugin, you can use this helper function to get a handler for a specific hook.
Here's how you can use it:

```typescript
// Vitest stuff
import { describe, it, expect } from "vitest";

import { getHandler } from '@stash-it/dev-tools';

// Your plugin.
import { createPlugin } from './your-plugin';

describe('YourPlugin', () => {
  it('does something on buildKey', async () => {
    const plugin = createPlugin();
    const key = "key";
    const handler = getHandler('buildKey', plugin);
    
    // Here is where you can test the handler that you created.
    const result = await handler(key);
    
    // expect ... etc.
  });
});
```

Why not simply do this?

```typescript
const handler = plugin.hookHandlers.buildKey;
```

Well ... current types set around plugins are not that great.
This helper function will throw an error if the handler is not a function. Also, thanks to the argument passed (hook name),
it will know what arguments the handler should accept etc. And you don't need to reach to plugin internals. You simply
request a handler for a specific hook and that's it.

## License

MIT

## Contribution

Feel free to open an issue or a pull request.
