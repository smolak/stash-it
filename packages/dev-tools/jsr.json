{
  "name": "@stash-it/dev-tools",
  "version": "0.0.0-beta.4",
  "exports": "./mod.ts",
  "publish": {
    "include": ["package.json", "README.md", "mod.ts", "src/**/*.ts"],
    "exclude": ["src/**/*.test.ts"]
  }
}
