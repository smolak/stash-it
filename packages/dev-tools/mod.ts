export * from "./src/test/adapter-test-cases";
export * from "./src/test/helpers";
