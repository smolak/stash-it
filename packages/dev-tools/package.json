{
  "name": "@stash-it/dev-tools",
  "version": "0.0.5",
  "type": "module",
  "publishConfig": {
    "access": "public"
  },
  "exports": "./mod.ts",
  "files": [
    "mod.ts",
    "src",
    "_dist",
    "README.md",
    "package.json"
  ],
  "license": "MIT",
  "keywords": [
    "@stash-it",
    "stash it",
    "stash-it",
    "dev-tools",
    "tools"
  ],
  "scripts": {
    "lint": "eslint .",
    "publish": "pnpx jsr publish"
  },
  "dependencies": {
    "@stash-it/core": "npm:@jsr/stash-it__core@0.3.0",
    "nanoid": "^5.0.8",
    "vitest": "^2.1.4"
  }
}
