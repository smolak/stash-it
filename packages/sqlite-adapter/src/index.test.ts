import fs from "fs-extra";
import path from "path";
import { it, expect, afterAll, describe } from "vitest";
import SqliteDatabase from "better-sqlite3";
import { runAdapterTests } from "@stash-it/dev-tools";
import { createHmac } from "node:crypto";

import { SqliteAdapter, type SqliteAdapterOptions } from "./index";

const generateRandomString = () => createHmac("sha256", "stash-it").digest("hex");

const tempDir = path.join(__dirname, "..", "temp");
const tableName = "items";
const keyColumnName = "key";
const valueColumnName = "value";
const extraColumnName = "extra";

let adapter: SqliteAdapter;

const dbName = `testdb_${generateRandomString()}.db`;
const dbPath = path.join(tempDir, dbName);

const createAdapter = (options: Partial<SqliteAdapterOptions> = {}) => {
  const db = new SqliteDatabase(dbPath);

  db.prepare(
    `CREATE TABLE ${tableName} (${keyColumnName} TEXT PRIMARY KEY, ${valueColumnName} TEXT, ${extraColumnName} TEXT)`,
  ).run();

  return new SqliteAdapter({ ...options, dbPath });
};

fs.removeSync(tempDir);
fs.ensureDirSync(tempDir);

adapter = createAdapter();

describe("sqlite-adapter", () => {
  afterAll(() => {
    fs.removeSync(tempDir);
  });

  describe("validation", () => {
    it("should throw when db file does not exist", () => {
      expect(() => new SqliteAdapter({ dbPath: "./non-existent.db" })).to.throw("unable to open database file");
    });

    it("should throw when table does not exist", () => {
      expect(() => new SqliteAdapter({ dbPath, tableName: "non_existent_table" })).to.toThrowError(
        "no such table: non_existent_table",
      );
    });

    it("should throw when key column doesn't exist", () => {
      expect(() => new SqliteAdapter({ dbPath, keyColumnName: "non_existent_key_column" })).to.toThrowError(
        "no such column: non_existent_key_column",
      );
    });

    it("should throw when value column doesn't exist", () => {
      expect(() => new SqliteAdapter({ dbPath, valueColumnName: "non_existent_value_column" })).to.toThrowError(
        "no such column: non_existent_value_column",
      );
    });

    it("should throw when extra column doesn't exist", () => {
      expect(() => new SqliteAdapter({ dbPath, extraColumnName: "non_existent_extra_column" })).to.toThrowError(
        "no such column: non_existent_extra_column",
      );
    });

    it("should not throw when validation passes", () => {
      expect(() => new SqliteAdapter({ dbPath })).not.toThrow();
    });

    describe("optional parameters", () => {
      describe("tableName", () => {
        it("should throw if passed value is not a string", () => {
          // @ts-ignore
          expect(() => new SqliteAdapter({ dbPath, tableName: 1 })).toThrowErrorMatchingSnapshot();
        });
      });

      describe("keyColumnName", () => {
        it("should throw if passed value is not a string", () => {
          // @ts-ignore
          expect(() => new SqliteAdapter({ dbPath, keyColumnName: 1 })).toThrowErrorMatchingSnapshot();
        });
      });

      describe("valueColumnName", () => {
        it("should throw if passed value is not a string", () => {
          // @ts-ignore
          expect(() => new SqliteAdapter({ dbPath, valueColumnName: 1 })).toThrowErrorMatchingSnapshot();
        });
      });

      describe("extraColumnName", () => {
        it("should throw if passed value is not a string", () => {
          // @ts-ignore
          expect(() => new SqliteAdapter({ dbPath, extraColumnName: 1 })).toThrowErrorMatchingSnapshot();
        });
      });
    });
  });

  runAdapterTests(adapter);
});
