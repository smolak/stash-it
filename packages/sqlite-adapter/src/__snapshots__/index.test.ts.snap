// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sqlite-adapter > validation > optional parameters > extraColumnName > should throw if passed value is not a string 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "extraColumnName"
    ],
    "message": "Expected string, received number"
  }
]]
`;

exports[`sqlite-adapter > validation > optional parameters > keyColumnName > should throw if passed value is not a string 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "keyColumnName"
    ],
    "message": "Expected string, received number"
  }
]]
`;

exports[`sqlite-adapter > validation > optional parameters > tableName > should throw if passed value is not a string 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "tableName"
    ],
    "message": "Expected string, received number"
  }
]]
`;

exports[`sqlite-adapter > validation > optional parameters > valueColumnName > should throw if passed value is not a string 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "valueColumnName"
    ],
    "message": "Expected string, received number"
  }
]]
`;
